test/controllers/tasks_controller_test.rb

```diff
@@ -105,24 +107,137 @@
     # Note:  If there was a way to fail to save the changes to a task, that would be a great
     #        thing to test.
     it "can update an existing task" do
-      skip
-      # Your code here
+      # Arrange
+      test_id = Task.last.id
+      task_hash = {
+        task: {
+          name: "updated task",
+          description: "updated task description",
+        },
+      }
+
+      expect {
+        patch task_path(test_id), params: task_hash
+      }.wont_change "Task.count"
+
+      updated_task = Task.find_by(name: task_hash[:task][:name])
+
+      expect(updated_task.description).must_equal task_hash[:task][:description]
+
+      must_respond_with :redirect
+      must_redirect_to task_path(test_id)
     end
 
     it "will redirect to the root page if given an invalid id" do
-      skip
-      # Your code here
+      # Act
+      patch task_path(-1)
+
+      # Assert
+      must_respond_with :redirect
+      expect(flash[:error]).must_equal "Could not find task with id: -1"
+
+      must_redirect_to root_path
     end
   end
 
   # Complete these tests for Wave 4
   describe "destroy" do
-    # Your tests go here
+    it "removes the test from the database" do
+      # Arrange
+      test_id = Task.last.id
+
+      # Act
+      expect {
+        delete task_path(test_id)
+      }.must_change "Task.count", -1
+
+      # Assert
+      expect(Task.find_by(id: test_id)).must_equal nil
+
+      must_respond_with :redirect
+      must_redirect_to root_path
+    end
+
+    it "redirects to root if the book does not exist" do
+      # Arrange
+      test_id = -1
+
+      # Act
+      delete task_path(test_id)
 
+      # Assert
+      must_respond_with :redirect
+      expect(flash[:error]).must_equal "Could not find task with id: -1"
+
+      must_redirect_to root_path
+    end
   end
 
   # Complete for Wave 4
   describe "toggle_complete" do
-    # Your tests go here
+    it "can mark an incomplete task complete without changing anything else" do
+      # Arrange
+      test_task = Task.last
+      test_task.update completion_date: nil
+      initial_attributes = test_task.attributes.clone
+      task_hash = {
+        task: {
+          completion_date: Time.now.to_date,
+        },
+      }
+
+      #Act-Assert
+      expect {
+        patch toggle_complete_task_path(test_task.id), params: task_hash
```

Comment:
You don't need the params or the task_hash for this controller action. 

It is good that you are testing marking a complete and marking an incomplete task.